src/stream/xml_reader.rs

```diff
@@ -1,221 +1,262 @@
-            .trim_whitespace(false)
-            .whitespace_to_characters(true)
-            .cdata_to_characters(true)
-            .ignore_comments(true)
-            .coalesce_characters(true);
+        let mut xml_reader = EventReader::from_reader(BufReader::new(reader));
+        xml_reader.trim_text(true);
 
         XmlReader {
-            xml_reader: EventReader::new_with_config(reader, config),
-            queued_event: None,
-            element_stack: Vec::new(),
+            buffer: Vec::new(),
             finished: false,
+            state: ReaderState {
+                xml_reader,
+                closed_element: None,
+                element_stack: Vec::new(),
+            },
         }
     }
+}
 
-    fn read_content(&mut self) -> Result<String, Error> {
+impl From<XmlReaderError> for ErrorKind {
+    fn from(err: XmlReaderError) -> Self {
+        match err {
+            XmlReaderError::Io(err) if err.kind() == io::ErrorKind::UnexpectedEof => {
+                ErrorKind::UnexpectedEof
+            }
+            XmlReaderError::Io(err) => {
+                let err = if let Some(code) = err.raw_os_error() {
+                    io::Error::from_raw_os_error(code)
+                } else {
+                    io::Error::new(err.kind(), err.to_string())
+                };
+                ErrorKind::Io(err)
+            }
+            XmlReaderError::UnexpectedEof(_) => ErrorKind::UnexpectedEof,
+            XmlReaderError::Utf8(_) => ErrorKind::InvalidXmlUtf8,
+            _ => ErrorKind::InvalidXmlSyntax,
+        }
+    }
+}
+
+impl<R: Read> ReaderState<R> {
+    fn xml_reader_pos(&self) -> FilePosition {
+        let pos = self.xml_reader.buffer_position();
+        FilePosition(pos as u64)
+    }
+
+    fn with_pos(&self, kind: ErrorKind) -> Error {
+        kind.with_position(self.xml_reader_pos())
+    }
+
+    fn read_xml_event<'buf>(&mut self, buffer: &'buf mut Vec<u8>) -> Result<XmlEvent<'buf>, Error> {
+        let event = self.xml_reader.read_event(buffer);
+        let pos = self.xml_reader_pos();
+        event.map_err(|err| ErrorKind::from(err).with_position(pos))
+    }
+
+    fn read_content(&mut self, buffer: &mut Vec<u8>) -> Result<String, Error> {
         loop {
-            match self.xml_reader.next() {
-                Ok(XmlEvent::Characters(s)) => return Ok(s),
-                Ok(event @ XmlEvent::EndElement { .. }) => {
-                    self.queued_event = Some(event);
-                    return Ok("".to_owned());
-                }
-                Ok(XmlEvent::EndDocument) => {
-                    return Err(self.with_pos(ErrorKind::UnclosedXmlElement))
+            match self.read_xml_event(buffer)? {
+                XmlEvent::Text(text) => {
+                    let pos = self.xml_reader_pos();
+                    let unespcaped = text
+                        .unescaped()
+                        .map_err(|err| ErrorKind::from(err).with_position(pos))?;
```

Comment:
Does `.map_err(|e| self.with_pos(ErrorKind::from(e)))?;` work here?

Reply:
yep, it does!